src/kvstore/raftex/RaftPart.cpp

```diff
@@ -572,63 +591,83 @@ void RaftPart::processAppendLogResponses(
         VLOG(2) << idStr_ << numSucceeded
                 << " hosts have accepted the logs";
 
-        std::unique_ptr<LogIterator> walIt;
         {
             std::lock_guard<std::mutex> g(raftLock_);
+            if (status_ != Status::RUNNING) {
+                // The partition is not running
+                VLOG(2) << idStr_ << "The partition is stopped";
+                sendingPromise_.setValue(AppendLogResult::E_STOPPED);
+                cachingPromise_.setValue(AppendLogResult::E_STOPPED);
+                logs_.clear();
+                replicatingLogs_ = false;
+                return;
+            }
 
+            if (role_ != Role::LEADER) {
+                // Is not a leader any more
+                VLOG(2) << idStr_ << "The leader has changed";
+                sendingPromise_.setValue(AppendLogResult::E_NOT_A_LEADER);
+                cachingPromise_.setValue(AppendLogResult::E_NOT_A_LEADER);
+                logs_.clear();
+                replicatingLogs_ = false;
+                return;
+            }
+            if (currTerm != term_) {
+                LOG(INFO) << idStr_ << "The leader has changed, ABA problem.";
+                sendingPromise_.setValue(AppendLogResult::E_TERM_OUT_OF_DATE);
+                replicatingLogs_ = false;
+                return;
+            }
             lastLogId_ = lastLogId;
             lastLogTerm_ = currTerm;
 
             lastMsgSentDur_.reset();
 
-            walIt = wal_->iterator(committedId + 1, lastLogId);
-        }
-
-        // Step 3: Commit the batch
-        if (commitLogs(std::move(walIt))) {
-            std::lock_guard<std::mutex> g(raftLock_);
-
-            committedLogId_ = lastLogId;
+            auto walIt = wal_->iterator(committedId + 1, lastLogId);
+            // Step 3: Commit the batch
+            if (commitLogs(std::move(walIt))) {
```

Comment:
Good catch again 👍 

This absolutely is a potential problem. But I think we might want to take different solution to fix it. The reason I left ```commitLogs()``` out of the lock is because committing logs could take very long time. Putting it in the lock would block the normal write.

What about making WAL thread safe?

Reply:
Discussed with @dangleptr offline, he will remove the ```raftLock``` in the log append logic, so that appending logs will not be blocked by ```commitLogs()``` function call

Reply:
I have a question, if we remove the raftLock, how can we synchronize the multi thread to commit log?